### PR TITLE
Implement Objects on VWA

### PR DESCRIPTION
**Pull request by peterzhu2118**
# Variable Width Allocation: Objects

# Feature description

# Benchmark setup

Benchmarking was done on a bare-metal Ubuntu machine on AWS.

```
$ uname -a
Linux 5.13.0-1022-aws #24~20.04.1-Ubuntu SMP Thu Apr 7 22:10:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

glibc version:

```
$ ldd --version
ldd (Ubuntu GLIBC 2.31-0ubuntu9.2) 2.31
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

To measure memory usage over time, the [mstat tool](https://github.com/bpowers/mstat) was used.

master was benchmarked on commit [5ef3c7e](https://github.com/ruby/ruby/commit/5ef3c7ea2d1968c87f361b6615699b92cc6e5a9a). The branch was rebased on top of the same commit.

## railsbench

For railsbench, we ran the [railsbench benchmark](https://github.com/k0kubun/railsbench/blob/master/bin/bench). For both the performance and memory benchmarks, 25 runs were conducted for each branch.

There is a small improvement in RPS and response times with a small increase in max memory usage. We can see in the RSS over time graph that the memory behavior of the branch and master is very similar.

```
+-----------------------+--------+--------+-------------+
|                       | Branch | master | Improvement |
+-----------------------+--------+--------+-------------+
| RPS                   | 836.45 | 817.55 | 1.02x       |
| p50 (ms)              | 1.15   | 1.18   | 1.03x       |
| p90 (ms)              | 1.34   | 1.37   | 1.02x       |
| p99 (ms)              | 1.70   | 1.70   | 1.00x       |
| p100 (ms)             | 5.63   | 5.36   | 0.95x       |
| Max memory usage (MB) | 91.41  | 90.88  | 0.99x       |
+-----------------------+--------+--------+-------------+
```

railsbench

## rdoc generation

For rdoc generation, we ran the following script for the branch and master to collect memory usage.

```
for x in $(seq 25)
do
  sudo rm -rf .ext/rdoc; mstat -o mstat/branch_$x.tsv -freq 59 -- ruby --disable-gems "./libexec/rdoc" --root "." --encoding=UTF-8 --all --ri --op ".ext/rdoc" --page-dir "./doc" --no-force-update  "."
done
```

For rdoc generation, we see a decrease in memory usage when using Variable Width Allocation.

Average max memory usage for VWA: 302.04 MB

Average max memory usage for master: 314.19 MB

VWA uses 1.04x less memory.

rdoc

## Liquid benchmarks

We don't see a big change in performance in liquid benchmarks.

```
+----------------------+--------+--------+-------------+
|                      | Branch | master | Improvement |
+----------------------+--------+--------+-------------+
| Parse (i/s)          | 40.97  | 40.95  | 1.00x       |
| Render (i/s)         | 130.90 | 130.08 | 1.01x       |
| Parse & Render (i/s) | 29.30  | 29.26  | 1.00x       |
+----------------------+--------+--------+-------------+
```

## Microbenchmarks

These microbenchmarks are very favourable for VWA since the objects created 10 instance variables, so they are embedded in VWA and allocated on the malloc heap for master.

```
+----------+--------+--------+-------------+
|          | Branch | master | Improvement |
+----------+--------+--------+-------------+
| #ivar0   | 2.790k | 2.155k | 1.29x       |
| #ivar9   | 2.700k | 2.185k | 1.24x       |
| #ivar0=  | 2.523k | 1.999k | 1.26x       |
| #ivar9=  | 2.222k | 1.996k | 1.11x       |
+----------+--------+--------+-------------+
```

[comment]: <> (GitHub collapse format)
<details>
  <summary>Benchmark source code</summary>

```ruby
require "bundler/inline"
gemfile do
  source "https://rubygems.org"
  gem "benchmark-ips"
end

IVAR_COUNT = 10

class Foo
  IVAR_COUNT.times do |i|
    attr_accessor :"ivar#{i}"
  end

  def initialize
    IVAR_COUNT.times do |i|
      instance_variable_set(:"@ivar#{i}", nil)
    end
  end
end

Foo.new

COUNT = 10_000

objs = COUNT.times.map { Foo.new }

Benchmark.ips do |x|
  x.report("Foo#ivar0") do |times|
    i = 0
    while i < times
      objs.each { |o| o.ivar0 }
      i += 1
    end
  end

  x.report("Foo#ivar9") do |times|
    i = 0
    while i < times
      objs.each { |o| o.ivar9 }
      i += 1
    end
  end

  x.report("Foo#ivar0=") do |times|
    i = 0
    while i < times
      objs.each { |o| o.ivar0 = nil }
      i += 1
    end
  end

  x.report("Foo#ivar9=") do |times|
    i = 0
    while i < times
      objs.each { |o| o.ivar9 = nil }
      i += 1
    end
  end
end
```
</details>

### CODE REVIEW

**Comment by maximecb** [comment]
@peterzhu2118 if I understand corrrectly, you're enabling bigger objects in this PR with more embedded ivars. Can you explain the heuristic you use to size the objects when allocating them? Is it based on the number of ivars in the class?

For railsbench, are you using the original railsbench, or the `yjit-bench` version? The nice thing about the `yjit-bench` one is that it cuts out all the HTTP overhead to reduce noise and focus more directly on the ruby execution time. It's also easier to report an average time. I would encourage you to run the `yjit-bench` suite so you can get more benchmark results. That could help detect if there's a pathological case somewhere (unlikely but good to check).

File: include/ruby/internal/core/robject.h [hunk 1]
```diff
@@ -95,6 +97,14 @@ struct RObject {
     /** Basic part, including flags and class. */
     struct RBasic basic;
 
+#if USE_RVARGC
+    /**
+    * Number of instance variables.  This is per object; objects might
+    * differ in this field even if they have the identical classes.
+    */
+    uint32_t numiv;
+#endif
```
Comment (maximecb) [inline include/ruby/internal/core/robject.h:106]: I guess this is a matter of personal taste, but I find this code hard to parse. It might be more readable to have an `#if USE_RVARGC` that wraps the entire `RObject` struct, and two completely different struct definitions for the case with and without VWA.

But, IIUC, what you're doing here is making it so objects always have a `numiv` field, even when they are embedded.

**Comment by peterzhu2118** [comment]
I did not change the heuristic and it remains the same as before. It will eagerly allocate an ivar array with size equal to the number of ivars in the class. This is done in the [`rb_init_iv_list`](https://github.com/ruby/ruby/blob/71aec68566eb429e6c3cf86396bf2a22e383e080/variable.c#L1511) function that is called when the object is allocated.

For railsbench, I'm using [this script](https://github.com/k0kubun/railsbench/blob/master/bin/bench) which runs railsbench without performing any HTTP requests.

File: include/ruby/internal/core/robject.h (hunk 1 above)
Comment (peterzhu2118) [reply include/ruby/internal/core/robject.h:106]: The goal is to remove all non-VWA code in the future, so it's probably not here to stay for very long.

**Comment by peterzhu2118** [comment]
No I don't think that can be assumed right now, the following case can happen:

```ruby
class Foo
  def initialize(set_ivars)
    if set_ivars
      @ivar1 = nil
      @ivar2 = nil
      @ivar3 = nil
      @ivar4 = nil
    end
  end
end

a = Foo.new(false) # a has 0 ivars and space for 0 ivars
b = Foo.new(true) # b has 4 ivars and space for 4 ivars
c = Foo.new(false) # c has 0 ivars and space for 4 ivars
```

In this case, `a` does not have space for 4 ivars, but `c` will have space for 4 ivars.

**Comment by maximecb** [comment]
Ok so it's based on the ivar table and not on the number of ivars used in the class.

Is there a guaranteed minimum number of ivar slots available on objects? 🤔

File: gc.c [hunk 2]
```diff
@@ -2865,18 +2865,65 @@ rb_newobj(void)
...
+    }
+#else
+    size = sizeof(struct RObject);
+    if (index_tbl_num_entries > ROBJECT_EMBED_LEN_MAX) {
+        embed = false;
+    }
+#endif
+
+#if USE_RVARGC
+    VALUE obj = newobj_of(klass, flags, 0, 0, 0, wb_protected, size);
```
Comment (tenderlove) [inline gc.c:2899]: Why `0, 0, 0` instead of `Qundef`?  Just because we're iterating and manually setting to `Qundef` below?

File: include/ruby/internal/core/robject.h [hunk 3]
```diff
@@ -124,14 +135,30 @@ struct RObject {
             struct st_table *iv_index_tbl;
         } heap;
 
+#if USE_RVARGC
+        VALUE ary[1];
```
Comment (tenderlove) [inline include/ruby/internal/core/robject.h:139]: Can you add comments about this?  I don't understand what this field is for and why it's only `1`.

**Review by tenderlove — COMMENTED** [review]
It looks great!  I don't understand some of the changes so I left a comment

File: gc.c (hunk 2 above)
Comment (peterzhu2118) [reply gc.c:2899]: Yeah. Also [`rb_init_iv_list`](https://github.com/ruby/ruby/blob/8c1808151f4c1b44e8b0fe935c571f05b2641b8b/variable.c#L1515) reads the current number of ivars, and that needs to be 0 rather than `Qundef` (which is not 0).

**Comment by peterzhu2118** [comment]
@tenderlove Here's some benchmarks comparing this branch ("VWA") with this branch with VWA turned off ("VWA off") and the ruby_3_1 branch.

## railsbench

```
+-----------------------+--------+---------+--------+
|                       | VWA    | VWA off | 3.1    |
+-----------------------+--------+---------+--------+
| RPS                   | 788.38 | 768.68  | 727.67 |
| p50 (ms)              | 1.22   | 1.25    | 1.37   |
| p90 (ms)              | 1.42   | 1.40    | 1.43   |
| p99 (ms)              | 1.82   | 1.62    | 1.72   |
| p100 (ms)             | 6.01   | 5.01    | 15.46  |
| Max memory usage (MB) | 90.92  | 89.74   | 86.08  |
+-----------------------+--------+---------+--------+
```

railsbench

## rdoc generation

Average max memory usage for VWA on: 299.38 MB

Average max memory usage for VWA off: 353.35 MB

Average max memory usage for 3.1: 362.08 MB

rdoc

## Liquid benchmarks

```
+----------------------+---------+---------+---------+
|                      | VWA     | VWA off | 3.1     |
+----------------------+---------+---------+---------+
| Parse (i/s)          | 40.642  | 38.618  | 39.289  |
| Render (i/s)         | 127.810 | 118.789 | 122.313 |
| Parse & Render (i/s) | 28.958  | 27.318  | 27.763  |
| Memory usage (MB)    | 37.01   | 36.64   | 37.01   |
+----------------------+---------+---------+---------+
```
